src/lib.rs

```diff
@@ -248,31 +258,307 @@ impl<T, const N: usize> ArrayVec<T, { N }> {
+        let p = self.as_mut_ptr().add(index);
+        // Read the value before sending it to the other world.
+        let item = ptr::read(p);
+        // Shift every value after the removed one to the left.
+        ptr::copy(p.add(1), p, len - index - 1);
+        // We removed an item, so the length should be decremented.
+        self.set_len(len - 1);
+
+        item
+    }
+
+    /// Remove the value contained at `index` and return it without
+    /// conserving order.
+    ///
+    /// The removed value is replaced by the last value making this an
+    /// `O(1)` operation.
+    ///
+    /// # Panics
+    ///
+    /// The index is out of bounds.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use const_arrayvec::ArrayVec;
+    /// let mut vector = ArrayVec::from([1, 2, 4]);
+    ///
+    /// assert_eq!(vector.swap_remove(0), 1);
+    /// assert_eq!(&vector, [4, 2].as_ref());
+    ///
+    /// assert_eq!(vector.swap_remove(1), 2);
+    /// assert_eq!(&vector, [4].as_ref());
+    ///
+    /// assert_eq!(vector.swap_remove(0), 4);
+    /// assert_eq!(vector.len(), 0);
+    /// ```
+    pub fn swap_remove(&mut self, index: usize) -> T {
+        if let Some(item) = self.try_swap_remove(index) {
+            item
+        } else {
+            out_of_bounds!("swap_remove", index, self.len());
+        }
+    }
+
+    /// If the index is in bounds, remove the value contained at `index`
+    /// and return it without conserving order.
+    ///
+    /// The removed value is replaced by the last value making this an
+    /// `O(1)` operation.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use const_arrayvec::ArrayVec;
+    /// let mut vector = ArrayVec::from([1, 2, 4]);
+    ///
+    /// assert_eq!(vector.try_swap_remove(0), Some(1));
+    /// assert_eq!(&vector, [4, 2].as_ref());
+    ///
+    /// assert_eq!(vector.try_swap_remove(24), None);
+    /// assert_eq!(&vector, [4, 2].as_ref());
+    /// ```
+    #[inline]
+    pub fn try_swap_remove(&mut self, index: usize) -> Option<T> {
+        if index < self.len() {
+            Some(unsafe { self.swap_remove_unchecked(index) })
+        } else {
+            None
+        }
+    }
+
+    /// Remove the value contained at `index` and return it without
+    /// conserving order.
+    ///
+    /// The removed value is replaced by the last value making this an
+    /// `O(1)` operation.
+    ///
+    /// # Safety
+    ///
+    /// The index must be in bounds.
+    pub unsafe fn swap_remove_unchecked(&mut self, index: usize) -> T {
+        let new_len = self.len() - 1;
+        let p = self.as_mut_ptr();
+        let p_item = p.add(index);
+
+        // Read the item from its pointer.
+        let item = ptr::read(p_item);
+
+        // Check `index` isn't the last item's index.
+        if new_len != index {
+            // Replace the item at `index` with the last item.
+            ptr::copy_nonoverlapping(p.add(new_len), p_item, 1);
```

Comment:
I had to think for a minute here. What was the reasoning behind using `copy_nonoverlapping()` instead of something like

```rust
let p_last = p.add(new_len); 
*p_item = *p_last;
```

?

In most of the places I see `copy_nonoverlapping` used, it's for moving items around in bulk (i.e. arrays).

Reply:
You're... absolutely right. I guess I was so fixated on using the standard library that I forgot pointer dereferencing existed, it should also improve that function's performance too, thanks for noticing it!